File: app/src/main/java/io/lerk/lrkFM/tasks/archive/ArchiveUtil.java
package io.lerk.lrkFM.tasks.archive;

import android.net.IpSecManager;
import android.os.Looper;
import android.util.Log;

import com.github.junrar.Junrar;
import com.github.junrar.exception.RarException;

import org.apache.commons.compress.archivers.ArchiveException;
import org.apache.commons.compress.archivers.ArchiveInputStream;
import org.apache.commons.compress.archivers.ArchiveStreamFactory;
import org.apache.commons.compress.archivers.sevenz.SevenZArchiveEntry;
import org.apache.commons.compress.archivers.sevenz.SevenZFile;
import org.apache.commons.compress.archivers.zip.ZipArchiveEntry;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

import io.lerk.lrkFM.Handler;
import io.lerk.lrkFM.consts.FileType;
import io.lerk.lrkFM.entities.FMFile;
import io.lerk.lrkFM.exceptions.BlockingStuffOnMainThreadException;

/**
 * Archive utility class.
 *
 * @author Lukas Fülling (lukas@k40s.net)
 */
class ArchiveUtil {

    static {
        System.loadLibrary("sevenzipjbinding");
    }

    private static final String TAG = ArchiveUtil.class.getCanonicalName();

    boolean doExtractArchive(String destination, FMFile f) throws BlockingStuffOnMainThreadException {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            throw new BlockingStuffOnMainThreadException();
        }
        AtomicBoolean result = new AtomicBoolean(false);

        FileType fileType = f.getFileType();
        if (fileType != FileType.ARCHIVE_P7Z) {
            if(fileType == FileType.ARCHIVE_RAR) {
                fileType.newHandler(fmFile -> {
                    File archiveFile = fmFile.getFile();
                    File destinationFile = new File(destination);
                    if (!destinationFile.exists()) {
                        if (destinationFile.mkdirs()) {
                            Log.i(TAG, "Folder created: '" + destinationFile.getAbsolutePath() + "'");
                        } else {
                            Log.e(TAG, "Unable to create folder: '" + destinationFile.getAbsolutePath() + "'");
                        }
                    }
                    try {
                        Junrar.extract(archiveFile, destinationFile);
                        result.set(true);
                    } catch (RarException | IOException e) {
                        Log.e(TAG, "Unable to extract rar file '" + fmFile.getAbsolutePath() + "'!", e);
                        result.set(false);
                    }
                }).handle(f);
            } else {
                // handle archives using commons compress
                fileType.newHandler(getArchiveExtractionCallback(destination, result, fileType)).handle(f);
            }
        } else {
            fileType.newHandler(fi -> {
                SevenZFile sevenZFile;
                try {
                    sevenZFile = new SevenZFile(fi.getFile());
                    SevenZArchiveEntry entry;
                    while ((entry = sevenZFile.getNextEntry()) != null) {
                        if (entry.isDirectory()) {
                            continue;
                        }
                        File curfile = new File(destination, entry.getName());
                        File parent = curfile.getParentFile();
                        if (parent != null) {
                            if (!parent.exists()) {
                                if (parent.mkdirs()) {
                                    Log.i(TAG, "Folder created: '" + parent.getAbsolutePath() + "'");
                                } else {
                                    Log.e(TAG, "Unable to create folder: '" + parent.getAbsolutePath() + "'");
                                }
                            }
                            FileOutputStream out = new FileOutputStream(curfile);
                            byte[] content = new byte[(int) entry.getSize()];
                            sevenZFile.read(content, 0, content.length);
                            out.write(content);
                            out.close();
                        } else {
                            throw new IOException("Unable to get parent file for '" + curfile.getAbsolutePath() + "'!");
                        }
                    }
                    result.set(true);
                } catch (IOException e) {
                    Log.e(TAG, "Error extracting " + fileType.getExtension());
                    result.set(false);
                }
            }).handle(f);
        }
        return result.get();
    }

    private Handler<FMFile> getArchiveExtractionCallback(String path, AtomicBoolean result, FileType fileType) {
        return fi -> {
            try (InputStream is = new FileInputStream(fi.getFile())) {

                ArchiveInputStream ais = new ArchiveStreamFactory().createArchiveInputStream(fileType.getExtension(), is);
                ZipEntry entry;

                while ((entry = (ZipArchiveEntry) ais.getNextEntry()) != null) {

                    if (entry.getName().endsWith("/")) {
                        File dir = new File(path + File.separator + entry.getName());
                        if (!dir.exists()) {
                            //noinspection ResultOfMethodCallIgnored
                            dir.mkdirs();
                        }
                        continue;
                    }

                    File outFile = new File(path + File.separator + entry.getName());

                    if (outFile.isDirectory()) {
                        continue;
                    }

                    if (outFile.exists()) {
                        continue;
                    }

                    try (FileOutputStream out = new FileOutputStream(outFile)) {
                        byte[] buffer = new byte[1024];
                        //noinspection UnusedAssignment
                        int length = 0;
                        while ((length = ais.read(buffer)) > 0) {
                            out.write(buffer, 0, length);
                            out.flush();
                        }
                    }

                    result.set(true);
                }
            } catch (IOException | ArchiveException e) {
                Log.e(TAG, "Error extracting " + fileType.getExtension(), e);
                result.set(false);
            }

        };
    }

    boolean doCreateZip(CopyOnWriteArrayList<FMFile> files, File destination) throws BlockingStuffOnMainThreadException {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            throw new BlockingStuffOnMainThreadException();
        }
        try {
            FileOutputStream fos = new FileOutputStream(destination);
            ZipOutputStream zipOut = new ZipOutputStream(fos);
            for (FMFile f : files) {
                addFileToZip(f.getFile(), f.getName(), zipOut);
            }
            zipOut.close();
            fos.close();
            return true;
        } catch (IOException e) {
            Log.e(TAG, "unable to create zip file!", e);
        }
        return false;
    }

    /**
     * Adds a file to a ZipInputStream. Also walks subdirectories.
     *
     * @param f    the file
     * @param name the filename
     * @param zos  the ZipOutputStream
     * @throws IOException when there is an error
     */
    private static void addFileToZip(File f, String name, ZipOutputStream zos) throws IOException {
        if (f.isDirectory()) {
            File[] children = f.listFiles();
            for (File childFile : children) {
                addFileToZip(childFile, name + "/" + childFile.getName(), zos);
            }
            return;
        }
        FileInputStream fis = new FileInputStream(f);
        ZipEntry zipEntry = new ZipEntry(name);
        zos.putNextEntry(zipEntry);
        byte[] bytes = new byte[1024];
        int length;
        while ((length = fis.read(bytes)) >= 0) {
            zos.write(bytes, 0, length);
        }
        fis.close();
    }
}

File: app/src/main/java/io/lerk/lrkFM/tasks/archive/ArchiveUtil.kt
package io.lerk.lrkFM.tasks.archive

import android.os.Looper
import android.util.Log
import com.github.junrar.Junrar
import com.github.junrar.exception.RarException

import io.lerk.lrkFM.consts.FileType
import io.lerk.lrkFM.entities.FMFile
import io.lerk.lrkFM.exceptions.BlockingStuffOnMainThreadException
import org.apache.commons.compress.archivers.ArchiveEntry
import org.apache.commons.compress.archivers.ArchiveException
import org.apache.commons.compress.archivers.ArchiveStreamFactory
import org.apache.commons.compress.archivers.sevenz.SevenZArchiveEntry
import org.apache.commons.compress.archivers.sevenz.SevenZFile
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.atomic.AtomicBoolean
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream

/**
 * Archive utility class.
 *
 * @author Lukas Fülling (lukas@k40s.net)
 */
internal class ArchiveUtil {
    @Throws(BlockingStuffOnMainThreadException::class)
    fun doExtractArchive(destination: String?, f: FMFile?): Boolean {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            throw BlockingStuffOnMainThreadException()
        }
        val result = AtomicBoolean(false)
        val fileType = f?.fileType
        if (fileType != FileType.ARCHIVE_P7Z) {
            if (fileType == FileType.ARCHIVE_RAR) {
                fileType.newHandler { fmFile: FMFile ->
                    val archiveFile = fmFile.file
                    val destinationFile = File(destination!!)
                    if (!destinationFile.exists()) {
                        if (destinationFile.mkdirs()) {
                            Log.i(TAG, "Folder created: '" + destinationFile.absolutePath + "'")
                        } else {
                            Log.e(
                                TAG,
                                "Unable to create folder: '" + destinationFile.absolutePath + "'"
                            )
                        }
                    }
                    try {
                        Junrar.extract(archiveFile, destinationFile)
                        result.set(true)
                    } catch (e: RarException) {
                        Log.e(TAG, "Unable to extract rar file '" + fmFile.absolutePath + "'!", e)
                        result.set(false)
                    } catch (e: IOException) {
                        Log.e(TAG, "Unable to extract rar file '" + fmFile.absolutePath + "'!", e)
                        result.set(false)
                    }
                }.handle(f)
            } else {
                // handle archives using commons compress
                fileType!!.newHandler(getArchiveExtractionCallback(destination, result, fileType))
                    .handle(f)
            }
        } else {
            fileType.newHandler { fi: FMFile ->
                val sevenZFile: SevenZFile
                try {
                    sevenZFile = SevenZFile(fi.file)
                    var entry: SevenZArchiveEntry
                    while (sevenZFile.nextEntry.also { entry = it } != null) {
                        if (entry.isDirectory) {
                            continue
                        }
                        val curfile = File(destination, entry.name)
                        val parent = curfile.parentFile
                        if (parent != null) {
                            if (!parent.exists()) {
                                if (parent.mkdirs()) {
                                    Log.i(TAG, "Folder created: '" + parent.absolutePath + "'")
                                } else {
                                    Log.e(
                                        TAG,
                                        "Unable to create folder: '" + parent.absolutePath + "'"
                                    )
                                }
                            }
                            val out = FileOutputStream(curfile)
                            val content = ByteArray(entry.size.toInt())
                            sevenZFile.read(content, 0, content.size)
                            out.write(content)
                            out.close()
                        } else {
                            throw IOException("Unable to get parent file for '" + curfile.absolutePath + "'!")
                        }
                    }
                    result.set(true)
                } catch (e: IOException) {
                    Log.e(TAG, "Error extracting " + fileType.extension)
                    result.set(false)
                }
            }.handle(f)
        }
        return result.get()
    }

    private fun getArchiveExtractionCallback(
        path: String?,
        result: AtomicBoolean,
        fileType: FileType?
    ): (FMFile) -> Unit {
        return { fi: FMFile ->
            try {
                FileInputStream(fi.file).use { `is` ->
                    val ais = ArchiveStreamFactory().createArchiveInputStream(
                        fileType!!.extension,
                        `is`
                    )
                    var entry: ArchiveEntry
                    while (true) {
                        val nextEntry = ais.nextEntry ?: break
                        nextEntry.also { entry = it }
                        if (entry.name.endsWith("/")) {
                            val dir = File(path + File.separator + entry.name)
                            if (!dir.exists()) {
                                dir.mkdirs()
                            }
                            continue
                        }
                        val outFile = File(path + File.separator + entry.name)
                        if (outFile.isDirectory) {
                            continue
                        }
                        if (outFile.exists()) {
                            continue
                        }
                        FileOutputStream(outFile).use { out ->
                            val buffer = ByteArray(1024)
                            var length = 0
                            while (ais.read(buffer).also { length = it } > 0) {
                                out.write(buffer, 0, length)
                                out.flush()
                            }
                        }
                        result.set(true)
                    }
                }
            } catch (e: IOException) {
                Log.e(TAG, "Error extracting " + fileType!!.extension, e)
                result.set(false)
            } catch (e: ArchiveException) {
                Log.e(TAG, "Error extracting " + fileType!!.extension, e)
                result.set(false)
            }
        }
    }

    @Throws(BlockingStuffOnMainThreadException::class)
    fun doCreateZip(files: CopyOnWriteArrayList<FMFile?>?, destination: File?): Boolean {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            throw BlockingStuffOnMainThreadException()
        }
        try {
            val fos = FileOutputStream(destination)
            val zipOut = ZipOutputStream(fos)
            for (f in files!!) {
                addFileToZip(f!!.file, f.name, zipOut)
            }
            zipOut.close()
            fos.close()
            return true
        } catch (e: IOException) {
            Log.e(TAG, "unable to create zip file!", e)
        }
        return false
    }

    companion object {
        private val TAG = ArchiveUtil::class.java.canonicalName

        /**
         * Adds a file to a ZipInputStream. Also walks subdirectories.
         *
         * @param f    the file
         * @param name the filename
         * @param zos  the ZipOutputStream
         * @throws IOException when there is an error
         */
        @Throws(IOException::class)
        private fun addFileToZip(f: File?, name: String?, zos: ZipOutputStream) {
            if (f!!.isDirectory) {
                val children = f.listFiles()
                for (childFile in children!!) {
                    addFileToZip(childFile, name + "/" + childFile.name, zos)
                }
                return
            }
            val fis = FileInputStream(f)
            val zipEntry = ZipEntry(name)
            zos.putNextEntry(zipEntry)
            val bytes = ByteArray(1024)
            var length: Int
            while (fis.read(bytes).also { length = it } >= 0) {
                zos.write(bytes, 0, length)
            }
            fis.close()
        }
    }
}
